net/udpbroadcastrelay/main.c

```diff
@@ -0,0 +1,651 @@
+        }
+
+        /* Iterate through our interfaces and send packet to each one */
+        for (int iIf = 0; iIf < maxifs; iIf++) {
+            struct Iface* iface = &ifs[iIf];
+
+            /* no bounces, please */
+            if (iface == fromIface) {
+                continue;
+            }
+
+            struct in_addr fromAddress;
+            u_short fromPort;
+            if (spoof_addr == inet_addr("1.1.1.1")) {
+                fromAddress = iface->ifaddr;
+                fromPort = port;
+            } else if (spoof_addr == inet_addr("1.1.1.2")) {
+                fromAddress = iface->ifaddr;
+                fromPort = origFromPort;
+            } else if (spoof_addr) {
+                fromAddress.s_addr = spoof_addr;
+                fromPort = origFromPort;
+            } else {
+                fromAddress = origFromAddress;
+                fromPort = origFromPort;
+            }
+
+            struct in_addr toAddress;
+            if (rcv_inaddr.s_addr == INADDR_BROADCAST
+                || rcv_inaddr.s_addr == fromIface->dstaddr.s_addr) {
+                // Received on interface broadcast address -- rewrite to new interface broadcast addr
+                toAddress = iface->dstaddr;
+            } else {
+                // Send to whatever IP it was originally to
+                toAddress = rcv_inaddr;
+            }
+            u_short toPort = origToPort;
+
+            char fromAddressStr[255];
+            inet_ntoa2(fromAddress, fromAddressStr, sizeof(fromAddressStr));
+            char toAddressStr[255];
+            inet_ntoa2(toAddress, toAddressStr, sizeof(toAddressStr));
+            DPRINT (
+                "-> [ %s:%d -> %s:%d (iface=%d)\n",
+                fromAddressStr, fromPort,
+                toAddressStr, toPort,
+                iface->ifindex
+            );
+
+            /* Send the packet */
+            gram[8] = ttl;
+            memcpy(gram+12, &fromAddress.s_addr, 4);
+            memcpy(gram+16, &toAddress.s_addr, 4);
+            *(u_short*)(gram+20)=htons(fromPort);
+            *(u_short*)(gram+22)=htons(toPort);
+            #if (defined __FreeBSD__ && __FreeBSD__ <= 10) || defined __APPLE__
+            *(u_short*)(gram+24)=htons(UDPHEADER_LEN + len);
+            *(u_short*)(gram+2)=HEADER_LEN + len;
+            #else
+            *(u_short*)(gram+24)=htons(UDPHEADER_LEN + len);
+            *(u_short*)(gram+2)=htons(HEADER_LEN + len);
+            #endif
+            struct sockaddr_in sendAddr;
+            sendAddr.sin_family = AF_INET;
+            sendAddr.sin_port = htons(toPort);
+            sendAddr.sin_addr = toAddress;
+
+            if (sendto(
+                iface->raw_socket,
+                &gram,
+                HEADER_LEN+len,
+                0,
+                (struct sockaddr*)&sendAddr,
+                sizeof(sendAddr)
+            ) < 0) {
+                perror("sendto");
+            }
+        }
+        DPRINT ("\n");
+    }
+}

```

Comment:
the main function is far too big to be readable and understandable. It should be split up in smaller functions if possible.

Reply:
I'll think about that one, not my code to start with. :)

I've spent a lot more time on the plugin than this.